modal/partial_function.py

```diff
@@ -360,6 +356,62 @@ def wrapper(raw_f: Callable[..., Any]) -> _PartialFunction:
     return wrapper
 
 
+@typechecked
+def _web_server(
+    port: int,
+    *,
+    startup_timeout: float = 5.0,  # Maximum number of seconds to wait for the web server to start.
+    label: Optional[str] = None,  # Label for created endpoint. Final subdomain will be <workspace>--<label>.modal.run.
+    custom_domains: Optional[Iterable[str]] = None,  # Deploy this endpoint on a custom domain.
+) -> Callable[[Callable[..., Any]], _PartialFunction]:
+    """Decorator that registers an HTTP web server inside the container.
+
+    This is similar to `@asgi_app` and `@wsgi_app`, but it allows you to expose a full HTTP server
+    listening on a container port. This is useful for servers written in other languages like Rust,
+    as well as integrating with non-ASGI frameworks like aiohttp and Tornado.
+
+    **Usage:**
+
+    ```python
+    import subprocess
+
+    @stub.function()
+    @modal.web_server(8000)
+    def my_file_server():
+        subprocess.Popen("python -m http.server -d / 8000", shell=True)
```

Comment:
The system waits for something to bind to the specified port within `startup_timeout` seconds. Otherwise it displays an error.

Reply:
Dumb question but can you not launch a server through the `http.server` method in pure python? Kinda funny to run a sub process call that runs a Python module. Haven’t looked at what’s that’s doing though!

Reply:
this is just example code right? so it seems fine to use a subprocess

Reply:
This was mostly a question for my own understanding! But also nice for examples to demonstrate the simplest way of doing things, when possible.

Reply:
> The system waits for something ....

Yeh I know I was just saying we could have a TODO to go the extra mile and detect if that something is dead and will never bind the port!

Reply:
Good point @mwaskom! I didn't know this but it turns out that you can use `http.server` in pure Python, not as a CLI.

But the pure Python version is actually a little bit more complicated, since we would need to start a separate thread and write out `http.server.SimpleHTTPRequestHandler`. Also it's less similar to typical usage of `@web_server` for people interoping with non-Python servers! 😅

<img width="849" alt="image" src="https://github.com/modal-labs/modal-client/assets/7550632/d33e8ac9-1bf5-417f-b719-692f3baaeb2f">

Re: @thundergolfer good point, maybe we could introduce regular checks if the port can bind and exit the container if it's stopped working.

Reply:
Agreed — that makes for a less compelling example!

Reply:
I added a check for `ClientConnectorError`so that we stop fetching new inputs if the server has died.